File: sdk/python/src/fountain/team.py
"""Durable teammates, their conversations, and their schedules."""

from typing import Any, Dict, Iterator, List, Optional, Union

from .conversation import Conversation
from .http import HttpClient
from .resolve import Resolver
from .run import Run
from .sse import stream_path


class Team:
    def __init__(self, http: HttpClient, resolver: Resolver) -> None:
        self._http = http
        self._resolver = resolver
        self.schedules = TeamSchedules(http, resolver)

    def list(self) -> List[Dict[str, Any]]:
        return self._http.list("/api/team")

    def get(self, agent: str) -> Dict[str, Any]:
        return self._http.data("GET", "/api/team/%s" % self._agent_id(agent))

    def add(self, agent: str, **options: Any) -> Dict[str, Any]:
        return self._http.data(
            "POST", "/api/team", body=dict(options, agent_id=self._agent_id(agent))
        )

    def remove(self, agent: str) -> None:
        self._http.request("DELETE", "/api/team/%s" % self._agent_id(agent))

    def rename(self, agent: str, name: Optional[str]) -> Dict[str, Any]:
        return self._http.data(
            "PATCH", "/api/team/%s" % self._agent_id(agent), body={"name": name}
        )

    def message(
        self,
        agent: str,
        prompt: str,
        *,
        images: Optional[List[Dict[str, Any]]] = None,
        timeout: Optional[float] = None,
        collect_events: bool = False,
    ) -> Run:
        body: Dict[str, Any] = {"prompt": prompt}
        if images:
            body["images"] = images

        def plan() -> Any:
            agent_id = self._agent_id(agent)
            try:
                before = self._http.data("GET", "/api/team/%s" % agent_id)
            except Exception:
                before = None
            existing = None
            if isinstance(before, dict) and isinstance(
                before.get("conversation"), dict
            ):
                existing = before["conversation"].get("id")
            after, turn_number = 0, 1
            if existing:
                handle = Conversation(self._http, str(existing))
                after = handle.cursor()
                turn_number = handle.last_turn_number() + 1
            sent = self._http.request(
                "POST", "/api/team/%s/messages" % agent_id, body=body
            )
            conversation_id = (
                sent.get("conversation_id") if isinstance(sent, dict) else None
            )
            conversation_id = conversation_id or existing
            if not conversation_id:
                raise RuntimeError(
                    "POST /api/team/%s/messages returned no conversation id" % agent_id
                )
            if conversation_id != existing:
                after, turn_number = 0, 1
            record = self._http.data("GET", "/api/conversations/%s" % conversation_id)
            return record, turn_number, after

        return Run(self._http, plan, timeout=timeout, collect_events=collect_events)

    def conversation(self, agent: str) -> Conversation:
        teammate = self.get(agent)
        value = teammate.get("conversation")
        conversation_id = value.get("id") if isinstance(value, dict) else None
        if not conversation_id:
            raise RuntimeError(
                "%s has no conversation yet — send it a message first" % agent
            )
        return Conversation(self._http, str(conversation_id))

    def history(self, agent: str) -> List[Dict[str, Any]]:
        return self._http.list("/api/team/%s/conversations" % self._agent_id(agent))

    def fresh_conversation(self, agent: str) -> Dict[str, Any]:
        return self._http.data(
            "POST", "/api/team/%s/conversations" % self._agent_id(agent)
        )

    def stream(
        self,
        *,
        streams: Optional[Union[List[str], str]] = None,
        **options: Any,
    ) -> Iterator[Dict[str, Any]]:
        selected = ",".join(streams) if isinstance(streams, list) else streams
        options.setdefault("blocks", True)
        return stream_path(self._http, "/api/team/stream", streams=selected, **options)

    def _agent_id(self, agent: str) -> str:
        return str(self._resolver.resolve("/api/agents", "agent", agent)["id"])


class TeamSchedules:
    def __init__(self, http: HttpClient, resolver: Resolver) -> None:
        self._http = http
        self._resolver = resolver

    def list(self, agent: Optional[str] = None) -> List[Dict[str, Any]]:
        path = (
            "/api/team/schedules"
            if not agent
            else "/api/team/%s/schedules" % self._agent_id(agent)
        )
        return self._http.list(path)

    def get(self, agent: str, schedule_id: str) -> Dict[str, Any]:
        return self._http.data("GET", self._path(agent, schedule_id))

    def create(self, agent: str, input: Dict[str, Any]) -> Dict[str, Any]:
        return self._http.data("POST", self._path(agent), body=input)

    def update(
        self, agent: str, schedule_id: str, patch: Dict[str, Any]
    ) -> Dict[str, Any]:
        return self._http.data("PATCH", self._path(agent, schedule_id), body=patch)

    def delete(self, agent: str, schedule_id: str) -> None:
        self._http.request("DELETE", self._path(agent, schedule_id))

    def run(self, agent: str, schedule_id: str) -> Any:
        return self._http.request("POST", self._path(agent, schedule_id) + "/run")

    def _path(self, agent: str, schedule_id: Optional[str] = None) -> str:
        base = "/api/team/%s/schedules" % self._agent_id(agent)
        return "%s/%s" % (base, schedule_id) if schedule_id else base

    def _agent_id(self, agent: str) -> str:
        return str(self._resolver.resolve("/api/agents", "agent", agent)["id"])

File: sdk/elixir/lib/fountain/team.ex
defmodule Fountain.Team do
  @moduledoc "Durable teammates, their conversations, and message streams."
  alias Fountain.{Conversation, Error, HTTP, Resolver, Run, SSE}
  defstruct [:http, :resolver, :schedules]

  def new(http, resolver),
    do: %__MODULE__{
      http: http,
      resolver: resolver,
      schedules: struct(Fountain.TeamSchedules, http: http, resolver: resolver)
    }

  def list(value), do: HTTP.list(value.http, "/api/team")

  def get(value, agent),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do: HTTP.data(value.http, "GET", "/api/team/#{id}")
      )

  def add(value, agent, opts \\ %{}),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do:
          HTTP.data(value.http, "POST", "/api/team", body: Map.put(Map.new(opts), "agent_id", id))
      )

  def remove(value, agent),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do: void(HTTP.request(value.http, "DELETE", "/api/team/#{id}"))
      )

  def rename(value, agent, name),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do: HTTP.data(value.http, "PATCH", "/api/team/#{id}", body: %{"name" => name})
      )

  def message(value, agent, prompt, opts \\ []) do
    body = %{"prompt" => prompt} |> optional("images", opts[:images])

    Run.new(
      value.http,
      fn ->
        id = agent_id!(value, agent)

        before =
          case HTTP.data(value.http, "GET", "/api/team/#{id}") do
            {:ok, item} -> item
            _ -> nil
          end

        existing = get_in(before || %{}, ["conversation", "id"])

        {after_cursor, turn_number} =
          if existing do
            conversation = Conversation.new(value.http, existing)

            {ok!(Conversation.cursor(conversation)),
             ok!(Conversation.last_turn_number(conversation)) + 1}
          else
            {0, 1}
          end

        sent = HTTP.request!(value.http, "POST", "/api/team/#{id}/messages", body: body)
        conversation_id = if(is_map(sent), do: sent["conversation_id"]) || existing

        if is_nil(conversation_id),
          do:
            raise(%Error{
              message: "POST /api/team/#{id}/messages returned no conversation id",
              kind: :api
            })

        {after_cursor, turn_number} =
          if conversation_id == existing, do: {after_cursor, turn_number}, else: {0, 1}

        conversation = HTTP.data!(value.http, "GET", "/api/conversations/#{conversation_id}")
        {conversation, turn_number, after_cursor}
      end,
      opts
    )
  end

  def conversation(value, agent) do
    with {:ok, teammate} <- get(value, agent) do
      case get_in(teammate, ["conversation", "id"]) do
        nil ->
          {:error,
           %Error{
             message: "#{agent} has no conversation yet — send it a message first",
             kind: :not_found
           }}

        id ->
          {:ok, Conversation.new(value.http, id)}
      end
    end
  end

  def history(value, agent),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do: HTTP.list(value.http, "/api/team/#{id}/conversations")
      )

  def fresh_conversation(value, agent),
    do:
      with(
        {:ok, id} <- agent_id(value, agent),
        do: HTTP.data(value.http, "POST", "/api/team/#{id}/conversations")
      )

  def stream(value, opts \\ []),
    do: SSE.stream_path(value.http, "/api/team/stream", Keyword.put_new(opts, :blocks, true))

  defp agent_id(value, agent) do
    with {:ok, item} <- Resolver.resolve(value.resolver, "/api/agents", "agent", agent),
         do: {:ok, item["id"]}
  end

  defp agent_id!(value, agent), do: ok!(agent_id(value, agent))
  defp ok!({:ok, value}), do: value
  defp ok!({:error, error}), do: raise(error)
  defp optional(map, _key, nil), do: map
  defp optional(map, _key, []), do: map
  defp optional(map, key, value), do: Map.put(map, key, value)
  defp void({:ok, _}), do: :ok
  defp void(error), do: error
end

defmodule Fountain.TeamSchedules do
  @moduledoc "Cron routines attached to teammates."
  alias Fountain.{HTTP, Resolver}
  defstruct [:http, :resolver]

  def list(value, agent \\ nil),
    do:
      if(agent,
        do:
          with(
            {:ok, id} <- agent_id(value, agent),
            do: HTTP.list(value.http, "/api/team/#{id}/schedules")
          ),
        else: HTTP.list(value.http, "/api/team/schedules")
      )

  def get(value, agent, id),
    do: with({:ok, path} <- path(value, agent, id), do: HTTP.data(value.http, "GET", path))

  def create(value, agent, input),
    do:
      with(
        {:ok, path} <- path(value, agent),
        do: HTTP.data(value.http, "POST", path, body: input)
      )

  def update(value, agent, id, patch),
    do:
      with(
        {:ok, path} <- path(value, agent, id),
        do: HTTP.data(value.http, "PATCH", path, body: patch)
      )

  def delete(value, agent, id),
    do:
      with(
        {:ok, path} <- path(value, agent, id),
        do: void(HTTP.request(value.http, "DELETE", path))
      )

  def run(value, agent, id),
    do:
      with(
        {:ok, path} <- path(value, agent, id),
        do: HTTP.request(value.http, "POST", path <> "/run")
      )

  defp path(value, agent, id \\ nil),
    do:
      with(
        {:ok, agent_id} <- agent_id(value, agent),
        do: {:ok, "/api/team/#{agent_id}/schedules" <> if(id, do: "/#{id}", else: "")}
      )

  defp agent_id(value, agent) do
    with {:ok, item} <- Resolver.resolve(value.resolver, "/api/agents", "agent", agent),
         do: {:ok, item["id"]}
  end

  defp void({:ok, _}), do: :ok
  defp void(error), do: error
end
